### PR TITLE
Use train for AWS connection

### PR DESCRIPTION
**Pull request by clintoncwolfe**
This PR does three related things:

1. Modifies the resources' AWS-facing backends to rely on the train-provided connection, enabling using the `-t aws://REGION/PROFILE ` target.
2. In support of the first item, splits AwsResourceMixin into AwsSingularResourceMixin and AwsPluralResourceMixin.  All plural resources are standardized on using the new Plural mixin.
3. One resource, aws_iam_access_key, was found not to be adaptable at all to the train connector system.  It was re-written to match typical resources, and its documentation expanded.

We now have a train release, 1.0, which provides the needed funtionality.  However, our inspec core still relies on train 0.32, so you'll need to finagle your local Gemfile to test.

If you have clones of inspec (current master), inspec-aws, and train (current master) as siblings, you can edit inspec-aws's Gemfile to reflect:
```
gem 'inspec', path: '../inspec'
gem 'train', path: '../train' 
```
You'll also need to lie to Bundler about train's version, by editing ../train/lib/version to say '0.32'

We don't want to merge this to master yet (ever?) because it will break things badly until inspec has the new train dependency.

Related: chef/train#229

Closes: #178

### CODE REVIEW

File: libraries/aws_cloudtrail_trail.rb [hunk 1]
```diff
@@ -65,11 +67,13 @@ def fetch_from_aws
...
   class Backend
-    class AwsClientApi
-      BackendFactory.set_default_backend(self)
+    class AwsClientApi < AwsBackendBase
+      AwsCloudTrailTrail::BackendFactory.set_default_backend(self)
+      self.aws_client_class = Aws::CloudTrail::Client
 
       def describe_trails(query)
-        AWSConnection.new.cloudtrail_client.describe_trails(query)
+        # aws_transport.aws_client(Aws::CloudTrail::Client).describe_trails(query)
```
Comment (jquick) [inline libraries/aws_cloudtrail_trail.rb]: Just a reminder to remove this.

File: libraries/aws_cloudtrail_trails.rb [hunk 2]
```diff
@@ -9,37 +9,39 @@ class AwsCloudTrailTrails < Inspec.resource(1)
     end
   '
```
Comment (jquick) [inline libraries/aws_cloudtrail_trails.rb]: should probably add `supports platform: 'aws'` here also (and any other resources that are missing it)

File: libraries/aws_ec2_security_groups.rb [hunk 3]
```diff
@@ -15,63 +15,42 @@ class AwsEc2SecurityGroups < Inspec.resource(1)
...
-  def fetch_from_backend(criteria)
+  def fetch_from_api
     @table = []
-    backend = AwsEc2SecurityGroups::BackendFactory.create
+    backend = AwsEc2SecurityGroups::BackendFactory.create(inspec_runner)
     # Note: should we ever implement server-side filtering
     # (and this is a very good resource for that),
     # we will need to reformat the criteria we are sending to AWS.
-    backend.describe_security_groups(criteria).security_groups.each do |sg_info|
+    backend.describe_security_groups({}).security_groups.each do |sg_info|
```
Comment (jquick) [inline libraries/aws_ec2_security_groups.rb]: Why are we removing the criteria option here? Is there an issue with the new connection? or stability issue?

File: libraries/aws_iam_access_key.rb [hunk 4]
```diff
@@ -13,94 +13,94 @@ class AwsIamAccessKey < Inspec.resource(1)
...
 
-  def initialize(opts, decorator = IamClientDecorator.new)
-    @access_key = opts[:access_key]
-    @username = opts[:username]
-    @id = @access_key ? @access_key.access_key_id : opts[:id]
-
-    @decorator = decorator
-  end
+  include AwsSingularResourceMixin
+  attr_reader :username, :access_key_id, :status, :create_date
```
Comment (jquick) [inline libraries/aws_iam_access_key.rb]: Nit: I am fan of ordering these by alpha as it makes it easier to find things if they grow.

**Review by miah — APPROVED** [review]

File: libraries/aws_s3_bucket.rb [hunk 5]
```diff
@@ -38,7 +38,7 @@ def public?
   def has_access_logging_enabled?
     return unless @exists
     # This is simple enough to inline it.
-    !AwsS3Bucket::BackendFactory.create.get_bucket_logging(bucket: bucket_name).logging_enabled.nil?
+    !AwsS3Bucket::BackendFactory.create(inspec_runner).get_bucket_logging(bucket: bucket_name).logging_enabled.nil?
```
Comment (jquick) [inline libraries/aws_s3_bucket.rb]: should this just be `!BackendFactory.create(inspec_runner)`? Both work but I saw you removed the others.

**Review by jquick — COMMENTED** [review]
This is amazing! Great work @clintoncwolfe. I really like what you did with the new backend stuff. Very slick.

File: libraries/aws_cloudtrail_trails.rb (hunk 2 above)
Comment (clintoncwolfe) [reply libraries/aws_cloudtrail_trails.rb]: Yikes, thanks for the reminder.  Will do.

File: libraries/aws_s3_bucket.rb (hunk 5 above)
Comment (clintoncwolfe) [reply libraries/aws_s3_bucket.rb]: It's just more concise without the prefix.  I think I started removing the prefixes and got tired.  I can make them consistent.

File: libraries/aws_ec2_security_groups.rb (hunk 3 above)
Comment (clintoncwolfe) [reply libraries/aws_ec2_security_groups.rb]: Two reasons: 
1.  Using AwsPluralResourceMixin, we'd get our "criteria" / resource parameters from instance variables, not from the method argument.
2.  This method doesn't accept any resource params at this point (see line 42) and will validate for that.  So, it sends an empty hash.

**Review by jquick — APPROVED** [review]
Nice work!

**Comment by clintoncwolfe** [comment]
To avoid breaking anyone using the inspec-aws resource pack, I'll merge to `core-onramp`, and use that as the upstream over on inspec.

**Pull request merged**